### PR TITLE
Add monitor incident task to palo alto xdr plugin

### PR DESCRIPTION
**Pull request by jmurphy-R7**
### Description

Describe the proposed changes:

  - Add task to monitor incident events for palo alto cortex xdr

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [x] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Comment by jmurphy-R7** [comment]
(icon_venv) [BFS-MBP-5164] jmurphy ~/rapid7/insightconnect-plugins/plugins/palo_alto_cortex_xdr (wildcards/DCA-23913-xdr-task) icon-validate ./
[*] Validating plugin at ./

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
318
[*] Plugin failed validation! The following validation errors occurred:

Validator "RegenerationValidator" failed!
	Cause: Fatal: No binfile found in Python plugin.


----
[*] Total time elapsed: 2454.901ms

File: plugins/palo_alto_cortex_xdr/plugin.spec.yaml [hunk 1]
```diff
@@ -4,7 +4,7 @@ products: [insightconnect]
 name: palo_alto_cortex_xdr
 title: Palo Alto Cortex XDR
 description: Stop modern attacks with the industry's first extended detection and response platform that spans your endpoints, network and cloud data
-version: 2.0.0
+version: 3.0.0
```
Comment (jrose-r7) [inline plugins/palo_alto_cortex_xdr/plugin.spec.yaml:7]: Change to 2.1.0 because there are no backwards incompatible changes

**Review by jrose-r7 — CHANGES_REQUESTED** [review]

File: plugins/palo_alto_cortex_xdr/plugin.spec.yaml (hunk 1 above)
Comment (jmurphy-R7) [reply plugins/palo_alto_cortex_xdr/plugin.spec.yaml]: Done, thanks!

**Review by mrinehart-r7 — DISMISSED** [review]

**Comment by mberezin-r7** [comment]
> @jmurphy-R7 approved - but your PR is failing static analysis checks. PluginValidator and UnitTest are failing due to slashes in the branch name - we do not currently support that naming schema.

Looks good to me too aside from the branch name. Can we please run `git branch -M DCA-23913-xdr-task`? Will approve upon validator success.

**Comment by jmurphy-R7** [comment]
NOTE: The xdr docs also list `alert_sources` as a value you can filter on, I initially supported this filter as well however this field caused the XDR API to behave inconsistently. The docs specify you can provide a list of values to filter on this field, but when you provide more than one value, it always returns 0 events. If you provide one value and also try to filter on a different field, you will always get a http 500 status code in return. Due to this I have elected to not add it as an input.

validation results: 
```
[BFS-MBP-5164] jmurphy ~/rapid7/insightconnect-plugins/plugins/palo_alto_cortex_xdr (wildcards/DCA-23913-xdr-task) icon-validate ./
[*] Validating plugin at ./

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator UnapprovedKeywordsValidator
320
[*] Plugin successfully validated!

----
[*] Total time elapsed: 1776.014ms
```

New output from task with these changes, with no filters on the request:
```
Connect: Connecting...
time sorting field: modification_time, request filters: []
Retrieving Cortex XDR Incident events between 1609502400000 and 1631184495813.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 516
Finished retrieving 1 new incident events this iteration.
time sorting field: modification_time, request filters: []
Retrieving Cortex XDR Incident events between 1631184495813 and 1631184497462.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Finished retrieving 0 new incident events this iteration.
```

with filters on the request:
```
Connect: Connecting...
time sorting field: creation_time, request filters: [
    {
        "field": "status",
        "operator": "eq",
        "value": "new"
    },
    {
        "field": "description",
        "operator": "in",
        "value": [
            "'Behavioral Threat' generated by XDR Agent detected on host msedgewin10 involving user ieuser",
            "BYE"
        ]
    },
    {
        "field": "incident_id_list",
        "operator": "in",
        "value": [
            "1",
            "2"
        ]
    }
]
Retrieving Cortex XDR Incident events between 1609502400000 and 1631184453369.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 516
Finished retrieving 1 new incident events this iteration.
time sorting field: creation_time, request filters: [
    {
        "field": "status",
        "operator": "eq",
        "value": "new"
    },
    {
        "field": "description",
        "operator": "in",
        "value": [
            "'Behavioral Threat' generated by XDR Agent detected on host msedgewin10 involving user ieuser",
            "BYE"
        ]
    },
    {
        "field": "incident_id_list",
        "operator": "in",
        "value": [
            "1",
            "2"
        ]
    }
]
Retrieving Cortex XDR Incident events between 1631184453369 and 1631184455093.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Finished retrieving 0 new incident events this iteration.
```
Since this code change now touches the code being used in the `get_alerts` and `get_incidents` trigger I re-ran those tests too:
```
[BFS-MBP-5164] jmurphy ~/rapid7/insightconnect-plugins/plugins/palo_alto_cortex_xdr (wildcards/DCA-23913-xdr-task) icon-plugin run -R tests/get_alerts.json
INFO[0000] Running command:  docker run --rm -i rapid7/palo_alto_cortex_xdr:2.1.0 --debug run < tests/get_alerts.json
Connect: Connecting...
rapid7/Palo Alto Cortex XDR:2.1.0. Step name: get_alerts
Initializing Get Alerts trigger for the Palo Alto Cortex XDR plugin.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/alerts/get_alerts_multi_events/ HTTP/1.1" 200 67
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/alerts/get_alerts_multi_events/ HTTP/1.1" 200 67
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/alerts/get_alerts_multi_events/ HTTP/1.1" 200 67
```

```
[BFS-MBP-5164] jmurphy ~/rapid7/insightconnect-plugins/plugins/palo_alto_cortex_xdr (wildcards/DCA-23913-xdr-task) icon-plugin run -R tests/get_incidents.json
INFO[0000] Running command:  docker run --rm -i rapid7/palo_alto_cortex_xdr:2.1.0 --debug run < tests/get_incidents.json
Connect: Connecting...
rapid7/Palo Alto Cortex XDR:2.1.0. Step name: get_incidents
Initializing Get Incidents trigger for the Palo Alto Cortex XDR plugin.
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
Starting new HTTPS connection (1): api-rapid7.xdr.us.paloaltonetworks.com:443
https://api-rapid7.xdr.us.paloaltonetworks.com:443 "POST /public_api/v1/incidents/get_incidents/ HTTP/1.1" 200 70
```

**Comment by jmurphy-R7** [comment]
Renamed branch, new PR at: https://github.com/rapid7/insightconnect-plugins/pull/1002, closing this PR.

**Pull request closed without merging**